src/Microsoft.VisualStudio.Threading.Tests/AsyncReaderWriterLockTests.cs

```diff
@@ -1813,6 +1813,49 @@ public void WriteLockAsyncSynchronousReleaseAllowsOtherWriters()
             Assert.True(firstLockTask.Wait(TestTimeout)); // rethrow any exceptions
         }
 
+        [StaFact]
+        public void WriteLockDisposingShouldNotBlockByOtherWriters()
+        {
+            var firstLockToRelease = new AsyncManualResetEvent();
+            var firstLockAccquired = new AsyncManualResetEvent();
+            var firstLockToDispose = new AsyncManualResetEvent();
+            var firstLockTask = Task.Run(async delegate
+            {
+                using (var firstLock = await this.asyncLock.WriteLockAsync())
+                {
+                    firstLockAccquired.Set();
+                    await firstLockToRelease.WaitAsync();
+                    await firstLock.ReleaseAsync();
+                    await firstLockToDispose.WaitAsync();
+                }
+            });
+
+            var secondLockReleased = new AsyncManualResetEvent();
+
+            var secondLockTask = Task.Run(async delegate
+            {
+                await firstLockAccquired.WaitAsync();
+                var awaiter = this.asyncLock.WriteLockAsync().GetAwaiter();
+                Assert.False(awaiter.IsCompleted);
+                awaiter.OnCompleted(() =>
+                {
+                    firstLockToDispose.Set();
+
+                    // We must block the thread synchorizely, so it won't release the NonConcurrentSynchronizationContext.
+                    firstLockTask.Wait();
```

Comment:
You should add a timeout here so that this thread isn't deadlocked forever even after the overall test method times out and fails. Otherwise all subsequent tests will have this thread stuck and that could complicate investigations.

Reply:
The test will timeout, because we have one when we wait on that secondLockTask. But I guess it will leave the task there and not cleaned up.  I added a timeout here as well.